torch/csrc/Stream.cpp

```diff
@@ -256,6 +261,112 @@ static PyObject* THPStream_eq(THPStream* self, THPStream* other) {
+  auto ctx_stream =
+      py::reinterpret_steal<py::object>(THPStream_Wrap(cur_stream));
+  PyObject_SetAttrString(_self, "_ctx_device_index", ctx_device_index.ptr());
+  PyObject_SetAttrString(_self, "_ctx_stream", ctx_stream.ptr());
+  // Support with torch.Stream() as stream: works
+  Py_INCREF(_self);
+  return _self;
+  END_HANDLE_TH_ERRORS
+}
+
+static PyObject* THPStream_exit(PyObject* _self, PyObject* unused) {
+  HANDLE_TH_ERRORS
+  auto self = (THPStream*)_self;
+  // No operation is performed if the stream does not belong to an accelerator.
+  if (C10_UNLIKELY(!at::accelerator::isAccelerator(
+          static_cast<c10::DeviceType>(self->device_type)))) {
+    Py_RETURN_NONE;
+  }
+  auto ctx_stream = THPObjectPtr(PyObject_GetAttrString(_self, "_ctx_stream"));
+  auto ctx_device_index =
+      THPObjectPtr(PyObject_GetAttrString(_self, "_ctx_device_index"));
+  if ((!ctx_stream) || (!ctx_device_index)) {
+    throw python_error();
+  }
+  Py_DECREF(ctx_stream.get());
+  Py_DECREF(ctx_device_index.get());
+  THPStream* prev_stream = (THPStream*)ctx_stream.release();
+  auto prev_device_index =
+      THPUtils_unpackDeviceIndex(ctx_device_index.release());
+  at::accelerator::setCurrentStream(c10::Stream::unpack3(
+      prev_stream->stream_id,
+      static_cast<c10::DeviceIndex>(prev_stream->device_index),
+      static_cast<c10::DeviceType>(prev_stream->device_type)));
+  // Reset the current device to the previous device if they differ.
+  if (static_cast<c10::DeviceIndex>(self->device_index) != prev_device_index) {
+    at::accelerator::setDeviceIndex(prev_device_index);
+  }
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+
+static void THPStream_maybe_init_context(THPStream* self) {
+  if (!(self->context)) {
+    auto dict = THPObjectPtr(PyDict_New());
+    if (!dict) {
+      throw python_error();
+    }
+    self->context = dict.release();
+  }
+}
+
+static PyObject* THPStream_getattro(PyObject* _self, PyObject* attr_name) {
+  HANDLE_TH_ERRORS
+  auto self = (THPStream*)_self;
+  auto attr_name_str = THPUtils_unpackString(attr_name);
+  if (attr_name_str == "_ctx_device_index" || attr_name_str == "_ctx_stream") {
+    THPStream_maybe_init_context(self);
+    // borrowed reference
+    auto value = THPObjectPtr(PyDict_GetItem(self->context, attr_name));
+    Py_INCREF(value.get());
+    return value.release();
+  }
+  return PyObject_GenericGetAttr(_self, attr_name);
+  END_HANDLE_TH_ERRORS
+}
+
+static int THPStream_setattro(
+    PyObject* _self,
+    PyObject* attr_name,
+    PyObject* value) {
+  HANDLE_TH_ERRORS
+  auto self = (THPStream*)_self;
+  auto attr_name_str = THPUtils_unpackString(attr_name);
+  if (attr_name_str == "_ctx_device_index" || attr_name_str == "_ctx_stream") {
+    THPStream_maybe_init_context(self);
+    return PyDict_SetItem(self->context, attr_name, value);
+  }
+  return PyObject_GenericSetAttr(_self, attr_name, value);
+  END_HANDLE_TH_ERRORS_RET(-1)
+}
```

Comment:
If we keep this as is, lets still simplify it a bit by removing these.
And in the enter/exit, use PyDict APIs directly on the context object.

Reply:
Yes, remove these make our code simpiler.